hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/DynamicConfigAddWanReplicationConfigCodec.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.protocol.codec;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.Generated;
+import com.hazelcast.client.impl.protocol.codec.builtin.*;
+import com.hazelcast.client.impl.protocol.codec.custom.*;
+
+import javax.annotation.Nullable;
+
+import static com.hazelcast.client.impl.protocol.ClientMessage.*;
+import static com.hazelcast.client.impl.protocol.codec.builtin.FixedSizeTypesCodec.*;
+
+/*
+ * This file is auto-generated by the Hazelcast Client Protocol Code Generator.
+ * To change this file, edit the templates or the protocol
+ * definitions on the https://github.com/hazelcast/hazelcast-client-protocol
+ * and regenerate it.
+ */
+
+/**
+ * Adds a WAN replication configuration.
+ */
+@Generated("b5eee93c147c31cdbf2c85e0098be718")
+public final class DynamicConfigAddWanReplicationConfigCodec {
+    //hex: 0x1B1200
+    public static final int REQUEST_MESSAGE_TYPE = 1774080;
+    //hex: 0x1B1201
+    public static final int RESPONSE_MESSAGE_TYPE = 1774081;
+    private static final int REQUEST_INITIAL_FRAME_SIZE = PARTITION_ID_FIELD_OFFSET + INT_SIZE_IN_BYTES;
+    private static final int RESPONSE_INITIAL_FRAME_SIZE = RESPONSE_BACKUP_ACKS_FIELD_OFFSET + BYTE_SIZE_IN_BYTES;
+
+    private DynamicConfigAddWanReplicationConfigCodec() {
+    }
+
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings({"URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD"})
+    public static class RequestParameters {
+
+        /**
+         * WAN replication configuration name.
+         */
+        public java.lang.String name;
+
+        /**
+         * A serialized {@link com.hazelcast.config.WanConsumerConfig} instance.
+         */
+        public @Nullable com.hazelcast.internal.serialization.Data consumerConfig;
```

Comment:
I started down the route of unrolling all the `*Config` types and defining them as custom types in `Custom.yaml`. I was looking at full rather than partial support: it seemed to break down in a few places, e.g. use of `Comparable` in `properties` and the many composite types in `WanBatchPublisherConfig`.

Reply:
@hazelcast/apis could you please advise?

Reply:
Yeah. we should create codecs for each them, and if there is a field like properties which uses Comparable, use Data just for that field. so that field should be a Map<String, Data>

Reply:
As an example, I checked for the `WanConsumerConfig` class. Unfortunately the design for the properties is made very flexible as if the user can provide any key as string and any `Comparable` value. I look at the example usages and most test cases we used String for the value type\ I do not know why it was made as Comparable. In any case, in current situation the value for the properties map entry seems to be a user provided object which we have no control on, it has to be Data, hence, for that field a Map<String, Data> would be suitable. Other fields, the `className` and `persistWanReplicatedData` can be encoded into the ClientMessage, but the implementation is again a user object hence, that should also be `Data`.

Reply:
Thanks @ihsandemir. I spoke with @srknzl about this yesterday and we discovered this 'holder' pattern that people were using as an intermediate state before being mapped to the final state (the existing configuration type). So, I am in the process of using such a pattern.

Reply:
Wow, this turned into a lot of effort due to boilerplate for mapping between protocol->holder->config. I updated the client protocol PR as well: https://github.com/hazelcast/hazelcast-client-protocol/pull/470/commits/52705f4e3d2fde5c3444cc3074c073529caaf38b